yjit_codegen.c

```diff
@@ -1323,10 +1314,45 @@ gen_jump(jitstate_t* jit, ctx_t* ctx)
     return YJIT_END_BLOCK;
 }
 
+// Guard that recv_opnd has the same class as known_klass. Recompile as contingency if possible, or take side exit a last resort.
+static bool
+jit_guard_known_klass(jitstate_t *jit, const ctx_t *recompile_context, VALUE known_klass, x86opnd_t recv_opnd, const int max_chain_depth, uint8_t *side_exit)
+{
+    // Can't guard for for these classes because some of they are sometimes immediate (special const).
+    // Can remove this by adding appropriate dynamic checks.
+    if (known_klass == rb_cInteger ||
+        known_klass == rb_cSymbol ||
+        known_klass == rb_cFloat ||
+        known_klass == rb_cNilClass ||
+        known_klass == rb_cTrueClass ||
+        known_klass == rb_cFalseClass) {
+        return false;
+    }
+
+    // Check that the receiver is a heap object
+    {
+        test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK));
+        jnz_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qfalse));
+        je_ptr(cb, side_exit);
+        cmp(cb, REG0, imm_opnd(Qnil));
+        je_ptr(cb, side_exit);
+    }
```

Comment:
Why do you insert blocks between curly braces like this one? There isn't a need for a new local scope here?

Reply:
Just for aesthetics. Do you prefer removing it?

Reply:
In the future yes, but I will merge this PR now :)